### ISSUE TITLE
"bad decrypt" while encrypting o.O

### ISSUE THREAD

**Issue by filiptrzeciak** [openssl/openssl#21439]
```
$ openssl version
OpenSSL 3.0.9 30 May 2023 (Library: OpenSSL 3.0.9 30 May 2023)

$ echo jebs | openssl enc -aes-256-cbc -pass pass:password -pbkdf2 -nopad
bad decrypt
40A72B209E7F0000:error:1C80006B:Provider routines:ossl_cipher_generic_block_final:wrong final block length:../providers/implementations/ciphers/ciphercommon.c:406:
```
Obviously blame -nopad, but to fail that way is unexpected. I'm just a user, so my expectations may vary. I hope this is usefull in any way.

**Comment by paulidale** [comment]
What else can be done except failing?  Without padding, the last block isn't complete.

What would your expectation be here?

**Comment by filiptrzeciak** [comment]
Idea of "decryption error" during encryption freaks me out. 

I was expecting padding input with zeros, despite -nopad , just for the sake of enc algorithm which can't go with random length streams on input.

On the other hand I could avoid trouble if manual mentioned that using -nopad may cause unhandled exception for block ciphers.

Thanks in advance for clarification.

**Comment by mattcaswell** [comment]
> Idea of "decryption error" during encryption freaks me out.

Yeah - that's a bug.

https://github.com/openssl/openssl/blob/5be8233d2be306a2906d3da16e59aa15a4559dd2/apps/enc.c#L691-L694

The above code needs changing to emit a "bad encrypt" message instead if `enc == 1`

**Comment by paulidale** [comment]
> I was expecting padding input with zeros, despite -nopad , just for the sake of enc algorithm which can't go with random length streams on input.

This would be padding which -nopad` is explicitly disabling.  Note that any such padding would be present on decryption -- i.e. what you started with wouldn't be identical to what you got after decryption.

Adding a note to the documentation might be reasonable.

**Comment by jaraco** [comment]
In jaraco/jaraco.crypto#5, I may have stumbled onto a different manifestation of this issue.

In that report, I've got one environment, my ARM macOS workstation running homebrew and OpenSSL 3.1.2 and 1.1.1v fails to _encrypt_ using the exact same routine that passes on Windows, Linux, and macOS in Github actions and Windows ARM locally.

I've put together this standalone repro:

```python
import ctypes
from ctypes import c_void_p, c_int, c_char, c_ulong

lib_path = '/opt/homebrew/opt/openssl@3/lib/libcrypto.dylib'
lib = ctypes.cdll.LoadLibrary(lib_path)

MAX_BLOCK_LENGTH = 32
MAX_IV_LENGTH = 16


class CipherType(ctypes.Structure):
    _fields_ = [
        ('nid', c_int),
        ('block_size', c_int),
        ('key_len', c_int),
        ('iv_len', c_int),
        ('flags', c_ulong),
        ('init', c_void_p),
        ('do_cipher', c_void_p),
        ('cleanup', c_void_p),
        ('ctx_size', c_int),
        ('set_asn1_parameters', c_void_p),
        ('get_asn1_parameters', c_void_p),
        ('ctrl', c_void_p),
        ('app_data', c_void_p),
    ]


class Cipher(ctypes.Structure):
    _fields_ = [
        ('cipher', c_void_p),  # POINTER(CipherType)
        ('engine', c_void_p),  # POINTER(ENGINE)
        ('encrypt', c_int),
        ('buf_len', c_int),
        ('oiv', c_char * MAX_IV_LENGTH),
        ('iv', c_char * MAX_IV_LENGTH),
        ('buf', c_char * MAX_BLOCK_LENGTH),
        ('num', c_int),
        ('app_data', c_void_p),
        ('key_len', c_int),
        ('flags', c_ulong),
        ('cipher_data', c_void_p),
        ('final_used', c_int),
        ('block_mask', c_int),
        ('final', c_char * MAX_BLOCK_LENGTH),
    ]


def get_error():
    err = lib.ERR_get_error()
    msg = ctypes.create_string_buffer(1024)
    lib.ERR_error_string(err, msg)
    return msg.value.decode()


lib.EVP_CIPHER_CTX_new.restype = ctypes.POINTER(Cipher)

ctx = lib.EVP_CIPHER_CTX_new().contents
assert ctx
key = '11111111111111111111111111111111'.encode()
iv = '1111111111111111'.encode()

engine = None
lib.EVP_get_cipherbyname.argtypes = (ctypes.c_char_p,)
lib.EVP_get_cipherbyname.restype = ctypes.POINTER(CipherType)
cipher = lib.EVP_get_cipherbyname('AES-256-CBC'.encode('ascii'))
assert cipher
encrypt = True
lib.EVP_CipherInit_ex.argtypes = (
    ctypes.POINTER(Cipher),
    ctypes.POINTER(CipherType),
    ctypes.c_void_p,
    ctypes.c_char_p,
    ctypes.c_char_p,
)

res = lib.EVP_CipherInit_ex(ctx, cipher.contents, engine, key, iv, encrypt)
if res == 0:
    print(f"Error initializing cipher: {get_error()}")
    raise SystemExit(1)

data = b''
out = ctypes.create_string_buffer(len(data) + MAX_BLOCK_LENGTH - 1)
out_len = ctypes.c_int()

out_data = []

lib.EVP_CipherUpdate.argtypes = (
    ctypes.POINTER(Cipher),
    ctypes.c_char_p,
    ctypes.POINTER(ctypes.c_int),
    ctypes.c_char_p,
    ctypes.c_int,
)

res = lib.EVP_CipherUpdate(ctx, out, out_len, data, len(data))
if res != 1:
    print(f"Error updating cipher: {get_error()}")
    raise SystemExit(1)

out_data.append(out.raw[: out_len.value])

out = ctypes.create_string_buffer(MAX_BLOCK_LENGTH)
out_len = ctypes.c_int()

lib.EVP_CipherFinal_ex.argtypes = (
    ctypes.POINTER(Cipher),
    ctypes.c_char_p,
    ctypes.POINTER(ctypes.c_int),
)
res = lib.EVP_CipherFinal_ex(ctx, out, out_len)
if res != 1:
    print(f"Error finalizing cipher: {get_error()}")
```

When run, it emits:

```
Error finalizing cipher: error:1C80006B:Provider routines::wrong final block length
```

When run on OpenSSL 1.1, it emits:

```
Error finalizing cipher: error:0606506D:digital envelope routines:EVP_DecryptFinal_ex:wrong final block length
```

Illustrating the "decrypt error during encrypt" behavior.

I tried turning on padding, but that didn't have any effect on the outcome.

It's currently hard-coded for openssl@3 on my mac, but you can change the library to a .so or .dll on Linux or Windows and it should work.

It's entirely possible I'm holding it wrong. Can someone help locate the fault?

**Comment by tom-cosgrove-arm** [comment]
@jaraco In [his comment above](https://github.com/openssl/openssl/issues/21439#issuecomment-1633937292), @mattcaswell clearly shows where the "bad decrypt" message is coming from. It is not the same as your issue, so could you please not piggy-back on this one?

I would suggest starting a discussion, rather than raising an issue, as currently your situation is more of a question ("what's happening") rather than a bug report ("there is this particular problem at this particular place in the code")

**Comment by jaraco** [comment]
https://github.com/openssl/openssl/discussions/21709

**Comment by trigpolynom** [comment]
Hi should this issue be closed?

**Comment by mattcaswell** [comment]
Fixed by #22134
